controllers/provider-gcp/pkg/webhook/controlplane/mutator.go

```diff
@@ -0,0 +1,242 @@
+// Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controlplane
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gardener/gardener-extensions/controllers/provider-gcp/pkg/gcp"
+	"github.com/gardener/gardener-extensions/controllers/provider-gcp/pkg/internal"
+	"github.com/gardener/gardener-extensions/pkg/webhook/controlplane"
+
+	"github.com/coreos/go-systemd/unit"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/operation/common"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	kubeletconfigv1beta1 "k8s.io/kubelet/config/v1beta1"
+)
+
+// newMutator creates a new controlplane mutator.
+func newMutator(unitSerializer controlplane.UnitSerializer, kubeletConfigCodec controlplane.KubeletConfigCodec, logger logr.Logger) *mutator {
```

Comment:
Why exactly? It's never used outside of this package, similarly to actuators, etc (which could be created by unexported constructors just as well). And of course, for the same reason it doesn't matter if the type it returns is the (exported) interface or the (unexported) concrete type.
In general, I am used to (and like) designs in which "building the dependency graph" (the code currently in `add.go`) belongs to a central place somewhere. In such designs, the constructors *must* be public and they *must* return the interface (since only the interface is exported). Do you envision a refactoring in this direction, or is there another reason behind this?

Reply:
You're using this inside your tests, and tests should not test private functions. We're also exporting all our `Actuator`s the same way in the packages although they're only used in one particular controller manager at the moment.

Reply:
The constructor code is not "tested", it's just used to create the object under test. Since the tests are in the same package, this is also not an issue. `Actuators` could just as well stay unexported.  
Anyway, I'll change it as you suggest - I just wanted to point out that I don't really see any benefit unless the code in `add.go` is moved to a different place.

Reply:
Done.